### PR TITLE
Jcg/ada 19 password reset token

### PR DESCRIPTION
**Pull request by jgeorge37**
(This is just for the backend, I have another story in the backlog to tie together all of the sign-in and password reset pieces)

This is to generate a password reset token - I.e. once everything comes together, the user can submit their email address to reset their password. When they do this, my API endpoint checks that the email address is associated with an account, and if so updates the reset_token column with a randomized string and generates an expiration timestamp (2 days in the future). This token will be used in the link that the user gets emailed to them. The token is a unique identifier for the user for the 2 days.

This change includes a database migration so even if you are not reviewing, once this gets merged you should npm run migrate on your local db.

Part 1 - To test the token generation:
0. Have a local DB (you know the drill) and something like Postman that is capable of doing post requests.
1. `npm run migrate` to update your db with the new columns
1. Add an account to your account table if your account table is empty (I would use psql for this personally)
2. `npm run dev`
3. Send a post request to http://localhost:3000/api/account/generate-token with the body containing "email": "emailThatYouHaveAnAccountFor@whatever.com" (but obvi with the actual email you made an account with in the account table)
4. Check that the token was created and that the expiration date is 2 days in the future (SELECT * FROM account;)
5. Try hitting the endpoint again with an email that has no associated account; you should get an relevant error message in the request response.

Part 2 - To test the password reset + token expiration:
0. `npm run dev`
1. Copy the generated token (prolly from psql) for the account you just did the above for.
2. Send a post request to http://localhost:3000/api/account/reset-pw with the body "token": THE TOKEN YOU COPIED and "password" THE NEW PASSWORD - dont keyboard smash bc in the db it will be encrypted so you have to remember it
3. Try the sign in page with the new password, you should not get a message about a bad password
4. To test expiration, set the time column for the account you have been messing with to the current time: UPDATE account SET token_expire_time = current_timestamp WHERE email = THE EMAIL OF THE ACCOOUNT
5. Try resetting the password again with Postman, using the same token. The response should contain a relevant error message.

and try to poke more holes if you want

### CODE REVIEW

**Review by cachemeifyoucan0 — COMMENTED** [review]
checked it out, added a test student and account with email, migrated, ran the app, but keep receiving an blank error responses - any tips on what I might be missing?
Screen Shot 2020-10-03 at 10 55 12 PM

**Comment by jgeorge37** [comment]
it might have something to do with how you have already run a migration 0007 locally. the console should provide an actual error message.

**Comment by cachemeifyoucan0** [comment]
> it might have something to do with how you have already run a migration 0007 locally. the console should provide an actual error message.

ok, I did revert all that before this but I'll see what I can find

**Comment by smiskus** [comment]
Oh I was also having this issue with the same error, so @cachemeifyoucan0 let me know if you find a solution

**Comment by jgeorge37** [comment]
@smiskus can you share the error from the console

**Comment by jgeorge37** [comment]
Is your inner env file pointed at heroku or your local db?

**Comment by cachemeifyoucan0** [comment]
> Is your inner env file pointed at heroku or your local db? @smiskus @cachemeifyoucan0

local - and actually the create account end point doesn't word for me either

**Comment by jgeorge37** [comment]
did you comment out the line in pg-query mentioned in the readme

**Comment by smiskus** [comment]
Yeah I commented out the pg-query line, only can see that it's a 500-Internal server error, and env file is set up for local. Also running into issues with doing npm run migrate so maybe that's causing it? Also attached .env file since I know I messed it up before: 
envFile

**Comment by jgeorge37** [comment]
Do you only have 1 env file? There should be 2

**Comment by cachemeifyoucan0** [comment]
> did you comment out the line in pg-query mentioned in the readme

FUK u right when i switch branches I stashed that and forgot - but now I'm getting a refusal 
Screen Shot 2020-10-03 at 11 51 02 PM

**Comment by jgeorge37** [comment]
What does your console say (terminal)?

**Comment by cachemeifyoucan0** [comment]
> What does your console say (terminal)?

HAH Im an idiot and didn't start the app again after that - ok but I do not see it updated in pdAdmin or by psql cli :(

**Comment by jgeorge37** [comment]
I mean what does your app console say like in VSCode or whatever and also can we switch to the slack dm w sara

File: acmw-db-app/pages/api/account/[pid].js [hunk 1]
```diff
@@ -20,6 +68,12 @@ export default async (req, res) => {
     
             if(pid === 'verify') {
                 result = await verify(body.email, body.password);
+            } else if(pid === 'generate-token') {
+                if(!body.email) throw ("Missing email address in request body.");
+                result = await generateToken(body.email);
+            } else if(pid === 'reset-pw') {
+                if(!body.password || !body.token) throw ("Missing token and/or new password");
+                result = await resetPassword(body.token, body.password);
```
Comment (SnobbyDragon) [inline acmw-db-app/pages/api/account/[pid].js:76]: I'm confused why does this only need token and password instead of also needing email?

**Review by SnobbyDragon — COMMENTED** [review]
This works awesomely :)

**Review by smiskus — APPROVED** [review]

**Pull request merged**